### ISSUE TITLE
branches

### ISSUE THREAD

**Issue by dvolgyes** [HazyResearch/learning-circuits#15]
It is a bit hard to follow which branch supposed to belong
to the published papers / blogposts.
e.g. this blog refers to the code, also mentions both 1x1 convolutions and fully connected layers:
https://dawn.cs.stanford.edu/2019/06/13/butterfly/

But the master branch (as far as i saw) only contains the replacement of nn.Linear,
and only branches, e.g. cnn branch contains convolutions.

Could you clarify which branch is the latest, or which one should be checked out?
(I understand that there are experiments, abandoned branches, etc.)

Is there any reason not to merge the latest work into the master branch?

**Comment by tridao** [comment]
Thanks for trying our code!

The `master` branch is the stable branch that should be used. It contains a drop-in replacement for nn.Linear. As mentioned in the blogpost, 1x1 conv is equivalent to matrix multiplication (i.e., nn.Linear) in the channel dimension, so one can write 1x1 conv as reshape and a linear layer:
```
input = torch.randn(b, c_in, h, w)  # Input of shape (batch, in_channel, height, width)
bfly = Butterfly(c_in, c_out, tied_weight=False)  # replacement for nn.Linear(c_in, c_out)
input_reshaped = input.view(b, c_in, h * w).transpose(1, 2).reshape(b * h * w, c_in)  # (batch * height * width, in_channel)
output = bfly(input_reshaped)  # (batch * height * width, out_channel)
output = output.view(b, h * w, c_out).transpose(1, 2).reshape(b, c_out, h, w)  # (batch, out_channel, height, width)
```
Thank you for your suggestion, we'll include this wrapper to make it easier to use 1x1 conv.

Regarding the `cnn` branch: we're trying out different experiments and different CUDA implementations that could be substantially faster. However, it's a bit unstable at the moment. We hope to merge it into master soon so that the butterfly multiplication is even faster.

Please let us know if you run into issues.

**Comment by dvolgyes** [comment]
Hi, 

Thank you for the explanation.
I would strongly argue for an alternative solution.
First of all, processing the whole batch makes a lots of complications.
You can easily transform it into a loop, and only process one sample at the time.
So your [B,C,...] problem would reduce to [C,...]
If it is first coordinate, no need for transposing, advanced indexing (strides) would be enough for sampling, and/or slicing / torch.unbind / torch.cat/torch.stack.
E.g. if you use unbind, you get a tuple of the layers, you can reorder elements in this tuple,
and create a new tensor with torch.stack.

My point is: transpose is very very expensive operation while slicing/indexing is very efficient.
With strides or torch.take you can make efficient subsampling and reordering.

If you insist on CUDA, go for it, but i think it is both overkill and unlikely to be faster than well implemented strides/indexing magic. (And i would also avoid it, because no offense, but custom cuda codes are above average error prone.)

One remark, which also could save quite some memory: if i understand correctly, you shuffle the channels in the butterfly convolution, add channels pairs together, and you shuffle them back. This makes sense when you calculate specific coefficients with specific meaning (e.g. Fourier). But if you just use butterfly for mixing channels, then the final order doesn't really matter, does it?
So cannot be this back-shuffling part omitted (and this way, saving half of the work)?

**Comment by tridao** [comment]
Thanks for the suggestion. 
I see that reshaping could add about 50% overhead (e.g. channel size 256 on GPU), since it requires copying the data.
We might eventually implement butterfly 1x1 conv directly in CUDA to avoid this overhead.

Regarding channel shuffling: it's not quite as simple as permuting the channels, adding pairs of them, then permuting them back. There are log(n) such steps to mix the channels (hence complexity is O(n log n) and not O(n)). In any case the expensive part isn't permuting the channels (and permuting them back), but rather the log(n) steps of adding pairs of channels together.

**Comment by dvolgyes** [comment]
I disagree in terms of what is expensive, memory access in CUDA is often way more expensive than the calculation itself, and transposing also has a terrible access pattern. Anyway, i will go for slicing, if needed, and good luck with the cuda version (but i recommend profiling the cuda code, so you actually optimize the bottleneck)
